### DISCUSSION TITLE
RLS - update statement - does id need to be included in data to be updated?

### DISCUSSION THREAD

**Discussion by tdeitz** [supabase/supabase#11958] [Questions] [answered]
Hello everyone :)

I have a client table, and using RLS I only want the user to be able to update the intake_completed field if their id matches the id field in the client table.

So the RLS clause would be: 
`  id = client.id`

My question is: do I need to include client id in the update call itself?

Specifically, will this work? 

```
const { error: intakeCompletedError } = await supabaseClient
          .from('client')
          .update({ intake_completed: true })
          .eq('id', clientId);
```

Or do I have to do: 

```
const { error: intakeCompletedError } = await supabaseClient
          .from('client')
          .update({ id: clientId, intake_completed: true })
          .eq('id', clientId);
```

(even though I don't actually need to update the id field itself) 

Thanks,

Tim

**Comment by GaryAustin1** [comment, marked as answer]
Update RLS USING looks at the entire row as it is before the update.  WITH CHECK looks at the entire row as it will be with the new data.

So you don't need to provide user_id.   I assume your RLS is `auth.uid() = client.id` or `auth.uid() = id`.   Not sure what `id=client.id` is if your table is client.

**Reply by tdeitz** [reply]
Thanks heaps for getting back to me Gary. Actually, I don't want to rely on `auth.uid()` in this instance. What is happening is that a clinician sends the client a survey link via email, the client completes the survey (without needing to log in), and the responses are used to update the client table. The **clientId** is extracted from the url (via the link), and used in the eventual db update statement: 
```

const { error: intakeCompletedError } = await supabaseClient
          .from('client')
          .update({  intake_completed: true })
          .eq('id', clientId);
```

So with the RLS
 `id = client.id`

I'm trying to say, _only update the row in the client table where the column **id** is equal to the **id** field being sent in the update statement._

Will that work?

Thanks again,

Tim

**Reply by GaryAustin1** [reply]
Ah.  You can't compare the current value to the new value in RLS directly.  
But can't you do this in an update trigger?  If new.id != old.id return null?

**Reply by tdeitz** [reply]
Thanks Gary.
I'm no expert when it comes to plpgsql:
I've just tried to create a new function update_only_matching_client_record: 

```
begin
    if(new.id != old.id) return null
  end;
```
However it won't save, saying I'm missing a THEN clause.

After saving the function, my plan was to create a trigger, to fire before the event, for an update operation on the client table.

Is this what you had in mind?

**Reply by GaryAustin1** [reply]
You have to have that in a trigger function to work, otherwise there is no new or old objects.
```
begin
   if (new.id != old.id) then return null;
   return new;
 end;
 ```
 Without checking....
 
And yes a before trigger on update and for each row

**Reply by GaryAustin1** [reply]
sorry it should be:
```
begin
   if (new.id != old.id) then return null;
   end if;
   return new;
 end;
 ```

**Reply by tdeitz** [reply]
Thanks heaps Gary, appreciate your help greatly. I was able to make the function and trigger. It's really on me to learn more about plpgsql. 

One last clarification - then I really will leave you in peace. Having created this function, to be triggered prior to any update of the **client** table — I can safely assume that this will DISALLOW the update statement if the id of the row trying to be updated doesn't match the id being passed in: 

i.e. so the below call would fail if clientId was some random uuid value that wasn't found in the client table?

```
const { error: intakeCompletedError } = await supabaseClient
          .from('client')
          .update({ intake_completed: true })
          .eq('id', clientId);
```

**Reply by GaryAustin1** [reply]
.eq has to match something in order for the update to occur.

But you still need some sort of RLS.  You are basically trusting the client to provide a proper filter.  Even if RLS is authenticated user I could modify the call to be .is('intake_completed', true) and then update all those columns to false.  I don't need to provide an id, they will match as they are not being changed.  A trigger or RLS does not know if a value was provided, just if the col is different.

If  your data does not depend on any RLS based on the jwt user id then I suspect what you are really looking for is an rpc call that REQUIRES a UUID to do the update.  So you would have RLS update policy FALSE.  And then an rpc function with the id as a parameter (like a not very secret secret.... all you have to do is know it).  The rpc function is set to "security definer" (yes more new stuff) to then bypass RLS, but it has to have the ID to do the sql update with it.   Then you really don't need to compare the new and old as you are forcing it in the rpc function.   It is either there so updates, or is not and fails.

I should have caught what you were trying to do earlier, but was just focused on the technical detail.

**Reply by tdeitz** [reply]
Thanks Gary, appreciate it :)
Tim

**Comment by tdeitz** [comment]
Thanks Gary, I'll marks this as answer - through actually, I still get a syntax error when I copy create a trigger function and copy this code in: 
```
begin
   if (new.id != old.id) then return null;
   return new;
 end;
```
I know I have to create the function first, then the trigger - but yeah, says there is a syntax error, "at or near ;"
